File: CpuA64/symbolic.cc

#include "CpuA64/symbolic.h"
#include <cstring>
#include "Util/assert.h"
#include "Util/parse.h"

namespace cwerg::a64 {
using namespace cwerg;

char* strappend(char* dst, std::string_view src) {
  std::memcpy(dst, src.data(), src.size());
  dst[src.size()] = 0;
  return dst + src.size();
}

char* strappenddec(char* dst, int64_t n) {
  if (n >= 0) {
    return dst + ToDecString(n, dst).size();
  } else {
    *dst++ = '-';
    return dst + ToDecString(-n, dst).size();
  }
}

char* strappendhex(char* dst, int64_t n) {
  dst = strappend(dst, "0x");
  return  dst + ToHexString(n, dst).size();
}

char* strappendflt(char* dst, double d) {
  return  dst + ToFltString(d, dst).size();
}

char* SymbolizeOperand(char* buf, uint32_t data, OK ok) {
  const FieldInfo& fi = FieldInfoTable[uint8_t(ok)];
  const uint64_t val = DecodeOperand(ok, data);
  switch (fi.kind) {
    default:
      ASSERT(false, "");
      return nullptr;
    case FK::LIST:
      return strappend(buf, fi.names[val]);
    case FK::FLT_CUSTOM: {
      return strappendflt(buf, Flt64FromBits(val));
    }
    case FK::INT_SIGNED:
    case FK::INT:  // TODO: add unsigned version
      return strappenddec(buf, (int64_t)val);
    case FK::INT_HEX_CUSTOM:
    case FK::INT_HEX:
      return strappendhex(buf, (int64_t)val);
  }
}

void SymbolizeReloc(char* cp, const Ins& ins, uint32_t addend) {
  ASSERT(false, "NYI");
}

std::string_view InsSymbolize(const a64::Ins& ins,
                              std::vector<std::string>* ops) {
  char buffer[128];
  for (unsigned i = 0; i < ins.opcode->num_fields; ++i) {
    if (ins.reloc_kind != elf::RELOC_TYPE_AARCH64::NONE && i == ins.reloc_pos) {
      SymbolizeReloc(buffer, ins, ins.operands[i]);
    } else {
      SymbolizeOperand(buffer, ins.operands[i], ins.opcode->fields[i]);
    }
    ops->emplace_back(buffer);
  }
  return ins.opcode->name;
}

uint32_t UnsymbolizeOperand(OK ok, std::string_view op) {
  const FieldInfo& fi = FieldInfoTable[uint8_t(ok)];
  uint64_t val;
  switch (fi.kind) {
    default:
      ASSERT(false, "");
      return kEncodeFailure;
    case FK::LIST:
      for (val = 0; val < fi.num_names; val++) {
        if (fi.names[val] == op) break;
      }
      if (val == fi.num_names) return kEncodeFailure;
      break;
    case FK::FLT_CUSTOM: {
      auto maybe = ParseFlt64(op);
      if (!maybe) return kEncodeFailure;
      val = Flt64ToBits(maybe.value());
      break;
    }
    case FK::INT_SIGNED: {
      auto maybe = ParseInt64(op);
      if (!maybe) return kEncodeFailure;
      val = maybe.value();
      break;
    }
    case FK::INT:  // TODO: add unsigned version
    case FK::INT_HEX_CUSTOM:
    case FK::INT_HEX: {
      auto maybe = ParseUint64(op);
      if (!maybe) return kEncodeFailure;
      val = maybe.value();
      break;
    }
  }

  return EncodeOperand(ok, val);
}

bool HandleRelocation(std::string_view expr, unsigned pos, Ins* ins) {
  ins->reloc_pos = pos;
  const size_t colon_sym = expr.find(':');
  if (colon_sym == std::string_view::npos) return false;
  const std::string_view kind_name = expr.substr(0, colon_sym);
  if (kind_name == "abs32") {
    ins->reloc_kind = elf::RELOC_TYPE_AARCH64::ABS32;
  } else if (kind_name == "abs64") {
      ins->reloc_kind = elf::RELOC_TYPE_AARCH64::ABS64;
  } else if (kind_name == "jump26") {
    ins->reloc_kind = elf::RELOC_TYPE_AARCH64::JUMP26;
    ins->is_local_sym = true;
  } else if (kind_name == "condbr19") {
    ins->reloc_kind = elf::RELOC_TYPE_AARCH64::CONDBR19;
    ins->is_local_sym = true;
  } else if (kind_name == "call26") {
    ins->reloc_kind = elf::RELOC_TYPE_AARCH64::CALL26;
  } else if (kind_name == "adr_prel_pg_hi21") {
    ins->reloc_kind = elf::RELOC_TYPE_AARCH64::ADR_PREL_PG_HI21;
  } else if (kind_name == "add_abs_lo12_nc") {
    ins->reloc_kind = elf::RELOC_TYPE_AARCH64::ADD_ABS_LO12_NC;
  } else if (kind_name == "loc_adr_prel_pg_hi21") {
    ins->reloc_kind = elf::RELOC_TYPE_AARCH64::ADR_PREL_PG_HI21;
    ins->is_local_sym = true;
  } else if (kind_name == "loc_add_abs_lo12_nc") {
    ins->reloc_kind = elf::RELOC_TYPE_AARCH64::TLSGD_ADD_LO12_NC;
    ins->is_local_sym = true;
  } else {
    return false;
  }
  //
  std::string_view rest = expr.substr(colon_sym + 1);
  const size_t colon_addend = rest.find(':');
  ins->reloc_symbol = rest.substr(0, colon_addend);

  ins->operands[pos] = 0;
  if (colon_addend != std::string_view::npos) {
    auto val = ParseInt<int32_t>(rest.substr(colon_addend + 1));
    if (!val.has_value()) return false;
    ins->operands[pos] = val.value();
  }
  return true;
}

bool InsFromSymbolized(const std::vector<std::string_view>& token, Ins* ins) {
  ins->opcode = FindOpcodeForMnemonic(token[0]);
  if (ins->opcode == nullptr) {
    std::cerr << "unknown opcode " << token[0] << "\n";
    return false;
  }
  ASSERT(token.size() - 1 == ins->opcode->num_fields,
         "bad number of token " << token.size() << " expected "
                                << ins->opcode->num_fields + 1 << " for "
                                << token[0]);
  for (unsigned i = 1; i < token.size(); ++i) {
    if (token[i].substr(0, 5) == "expr:") {
      if (!HandleRelocation(token[i].substr(5), i - 1, ins)) {
        std::cerr << "malformed relocation expression " << token[i] << "\n";
        return false;
      }
    } else {
      const OK ok = ins->opcode->fields[i - 1];
      auto val = UnsymbolizeOperand(ok, token[i]);
      if (val == kEncodeFailure) {
        std::cerr << "cannot parse " << token[i] << " for ok [" << int(ok)
                  << "]\n";
        return false;
      }
      ins->operands[i - 1] = val;
    }
  }
  return true;
}

// std::string_view InsSymbolize(const a64::Ins& ins,
//                              std::vector<std::string>* operands) {}

}  // namespace cwerg::a64

File: CpuA64/symbolic.py
"""
This module contains code for (un-)symbolizing the a64 ISA operands
"""
from typing import Any, Dict, List, Tuple
import struct

from Elf import enum_tab
from CpuA64 import opcode_tab as a64
from Util import  parse


def SymbolizeOperand(ok: a64.OK, data: int) -> str:
    t = a64.FIELD_DETAILS.get(ok)
    assert t is not None, f"NYI: {ok}"
    data = a64.DecodeOperand(ok, data)
    if t.kind == a64.FK.LIST:
        return t.names[data]
    elif t.kind == a64.FK.FLT_CUSTOM:
        # we only care about the float aspect
        data = parse.Flt64FromBits(data)
        return str(data)
    elif t.kind == a64.FK.INT_SIGNED:
        # we only care about the signed aspect
        if data >= 1 << 63:
            data -= (1 << 64)
        return str(data)
    elif t.kind == a64.FK.INT_HEX or t.kind == a64.FK.INT_HEX_CUSTOM:
        return hex(data)
    else:
        return str(data)


def _FloatTo64BitRepresentation(num: float) -> int:
    b = struct.pack('<d', num)
    assert len(b) == 8
    return int.from_bytes(b, "little")


def UnsymbolizeOperand(ok: a64.OK, op: str) -> int:
    """
    Converts a string into and int suitable for the provided `ok`
    E.g.
    #66 -> 65
    x0 -> 0
    asr -> 2
    """
    t = a64.FIELD_DETAILS.get(ok)
    assert t is not None, f"NYI: {ok}"

    if t.kind == a64.FK.LIST:
        # assert op in t, f"{op} not in {t}"
        data = t.names.index(op)
    elif t.kind == a64.FK.FLT_CUSTOM:
        # we only care about the float aspect
        data = parse.Flt64ToBits(float(op))
    else:
        data = int(op, 0)  # skip "#", must handle "0x" prefix
        # note we intentionally allow negative numbers here
    return a64.EncodeOperand(ok, data)


_RELOC_KIND_MAP = {
    # these relocations imply that the symbol is local
    "jump26": enum_tab.RELOC_TYPE_AARCH64.JUMP26,
    "condbr19": enum_tab.RELOC_TYPE_AARCH64.CONDBR19,
    # these relocations imply that the symbol is local
    # unless prefixed with `loc_`
    "call26": enum_tab.RELOC_TYPE_AARCH64.CALL26,
    "abs32": enum_tab.RELOC_TYPE_AARCH64.ABS32,
    "abs64": enum_tab.RELOC_TYPE_AARCH64.ABS64,

    "adr_prel_pg_hi21": enum_tab.RELOC_TYPE_AARCH64.ADR_PREL_PG_HI21,
    "add_abs_lo12_nc": enum_tab.RELOC_TYPE_AARCH64.ADD_ABS_LO12_NC,
}

_RELOC_OK = set([
    a64.OK.SIMM_PCREL_0_25, a64.OK.SIMM_PCREL_5_23, a64.OK.SIMM_PCREL_5_23_29_30
])


def _EmitReloc(ins: a64.Ins, pos: int) -> str:
    assert False, "NYI"


def InsSymbolize(ins: a64.Ins) -> Tuple[str, List[str]]:
    """Convert all the operands in an arm.Ins to strings including relocs
    """
    ops = []
    for pos, (ok, value) in enumerate(zip(ins.opcode.fields, ins.operands)):
        if (ok in _RELOC_OK and
                ins.reloc_kind != enum_tab.RELOC_TYPE_AARCH64.NONE and
                ins.reloc_pos == pos):
            ops.append(_EmitReloc(ins, pos))
        else:
            ops.append(SymbolizeOperand(ok, value))

    return ins.opcode.NameForEnum(), ops


def InsFromSymbolized(mnemonic: str, ops_str: List[str]) -> a64.Ins:
    opcode = a64.Opcode.name_to_opcode[mnemonic]
    ins = a64.Ins(opcode)
    for pos, (t, ok) in enumerate(zip(ops_str, opcode.fields)):
        if t.startswith("expr:"):
            # expr strings have the form expr:<rel-kind>:<symbol>:<addend>, e.g.:
            #   expr:movw_abs_nc:string_pointers:5
            #   expr:call:putchar
            rel_token = t.split(":")
            if len(rel_token) == 3:
                rel_token.append("0")
            assert rel_token[1] in _RELOC_KIND_MAP, f"unknown reloc kind {rel_token}"
            if rel_token[1] == "condbr19" or rel_token[1] == "jump26":
                ins.is_local_sym = True
            if rel_token[1].startswith("loc_"):
                ins.is_local_sym = True
                rel_token[1] = rel_token[1][4:]
            ins.reloc_kind = _RELOC_KIND_MAP[rel_token[1]]
            ins.reloc_pos = pos
            ins.reloc_symbol = rel_token[2]
            ins.operands.append(int(rel_token[3], 0))
        else:
            ins.operands.append(UnsymbolizeOperand(ok, t))
    return ins
